src/python/pants/build_graph/address_test.py

```diff
@@ -232,54 +231,65 @@ def test_address_equality() -> None:
-    assert normal_addr.spec == "a/b:c" == str(normal_addr) == normal_addr.reference()
-    assert normal_addr.path_safe_spec == "a.b.c"
-
-    top_level_addr = Address("", target_name="root")
-    assert top_level_addr.spec == "//:root" == str(top_level_addr) == top_level_addr.reference()
-    assert top_level_addr.path_safe_spec == ".root"
-
-    generated_addr = Address("a/b", relative_file_path="c.txt", target_name="c")
-    assert generated_addr.spec == "a/b/c.txt:c" == str(generated_addr) == generated_addr.reference()
-    assert generated_addr.path_safe_spec == "a.b.c.txt.c"
-
-    top_level_generated_addr = Address("", relative_file_path="root.txt", target_name="root")
-    assert (
-        top_level_generated_addr.spec
-        == "//root.txt:root"
-        == str(top_level_generated_addr)
-        == top_level_generated_addr.reference()
+    def assert_spec(address: Address, *, expected: str, expected_path_spec: str) -> None:
+        assert address.spec == expected
+        assert str(address) == expected
+        assert address.reference() == expected
+        assert address.path_safe_spec == expected_path_spec
+
+    assert_spec(Address("a/b", target_name="c"), expected="a/b:c", expected_path_spec="a.b.c")
+    assert_spec(Address("", target_name="root"), expected="//:root", expected_path_spec=".root")
+    assert_spec(
+        Address("a/b", relative_file_path="c.txt", target_name="c"),
+        expected="a/b/c.txt:c",
+        expected_path_spec="a.b.c.txt.c",
     )
-    assert top_level_generated_addr.path_safe_spec == ".root.txt.root"
-
-    generated_subdirectory_addr = Address(
-        "a/b", relative_file_path="subdir/c.txt", target_name="original"
+    assert_spec(
+        Address("", relative_file_path="root.txt", target_name="root"),
+        expected="//root.txt:root",
+        expected_path_spec=".root.txt.root",
+    )
+    assert_spec(
+        Address("a/b", relative_file_path="subdir/c.txt", target_name="original"),
+        expected="a/b/subdir/c.txt:../original",
+        expected_path_spec="a.b.subdir.c.txt@original",
     )
-    assert (
-        generated_subdirectory_addr.spec
-        == "a/b/subdir/c.txt:../original"
-        == str(generated_subdirectory_addr)
-        == generated_subdirectory_addr.reference()
+    assert_spec(
+        Address("a/b", relative_file_path="c.txt"),
+        expected="a/b/c.txt",
+        expected_path_spec="a.b.c.txt",
     )
-    assert generated_subdirectory_addr.path_safe_spec == "a.b.subdir.c.txt@original"
-
-    generated_addr_from_default_target = Address("a/b", relative_file_path="c.txt")
-    assert (
-        generated_addr_from_default_target.spec
-        == "a/b/c.txt"
-        == str(generated_addr_from_default_target)
-        == generated_addr_from_default_target.reference()
+    assert_spec(
+        Address("a/b", relative_file_path="subdir/f.txt"),
+        expected="a/b/subdir/f.txt:../b",
+        expected_path_spec="a.b.subdir.f.txt",
```

Comment:
This test was important: thanks.

The `path_safe_spec` should probably also include the target here though.